### ISSUE TITLE
Statusline overlaps windows in i3wm after #45.

### ISSUE THREAD

**Issue by alspore** [Lokaltog/candybar#47]
![2014-03-10-130609_1600x900_scrot](https://f.cloud.github.com/assets/1596161/2376410/49de8bf2-a876-11e3-89bd-393db85279ed.png)

**Comment by Lokaltog** [comment]
Did it work fine before #45? If so we may do some additional checks to see if the WM supports EWMH before we do the override redirect.

**Comment by Lokaltog** [comment]
It's a bit weird that this is happening since we're still setting `_NET_WM_STRUT_PARTIAL`, so the WM should reserve space on the top. I checked the i3 docs, which says the following:

> Regardless of which application you use to display the status line, you want to make sure that it registers as a dock window using EWMH hints. i3 will position the window either at the top or at the bottom of the screen, depending on which hint the application sets.

We do set the dock hint as well, but maybe the override-redirect hint overrides the dock hint so it stops working in i3.

**Comment by alspore** [comment]
![2014-03-10-140633_1600x900_scrot](https://f.cloud.github.com/assets/1596161/2377123/bcc39a9c-a87e-11e3-8b97-929eb9f89874.png)

This is a screenshot ~~post~~ pre issue #45.  My guess is that `_NET_WM_STRUT_PARTIAL` is being overridden.

**Comment by Lokaltog** [comment]
_Post_ or _pre_ issue #45?

If that's what it looked like before #45 then it looks like i3wm doesn't support `_NET_WM_STRUT_PARTIAL` at all, or that wkline conflicts with the other statusline that's running. What happens if you disable the i3 statusline?

**Comment by alspore** [comment]
Oops I made a type.  It's supposed to be pre issue.

**Comment by alspore** [comment]
This is what it looks like post #45 with i3bar disabled.
![2014-03-10-142420_1600x900_scrot](https://f.cloud.github.com/assets/1596161/2377315/48e9ec04-a881-11e3-8a04-a05b61fe858b.png)

**Comment by Lokaltog** [comment]
I'm guessing that i3wm doesn't support `_NET_WM_STRUT_PARTIAL` at all, and only cares about the docking hint. If it did support `_NET_WM_STRUT_PARTIAL` then it wouldn't be able to display two statuslines at the same time like on your screenshot since wkline only reserves space for itself.

I think we need to do a check to see what EWMH properties i3wm supports and not set the override-redirect hint, could you please post a gist with the output of `xprop -root` on your setup?

**Comment by alspore** [comment]
Here's the gist.  Apparently `_NET_WM_STRUT_PARTIAL` is supported.

https://gist.github.com/5paceManSpiff/9471153

**Comment by Lokaltog** [comment]
Hm. It says it supports `_NET_WM_STRUT_PARTIAL`. Maybe we could check for that property in `_NET_SUPPORTED` and don't set the override-redirect hint if it does support it.

**Comment by Lokaltog** [comment]
I've attempted to fix this by checking for EWMH/`_NET_WM_STRUT` support. If the required properties are supported then override-redirect won't be set.

@5paceManSpiff can you please check out commit c061d6930b1c04680ca712a7cf06f21252f458fd and see if this resolves the issue for you?

@bbenne10 can you please check if c061d6930b1c04680ca712a7cf06f21252f458fd still works in dwm/monsterwm?

**Comment by bbenne10** [comment]
Override_redirect will completely remove the window from view of the wm, so it makes sense that this is an issue. I'll check to make sure this still works in monsterwm tonight (dwm support is cool, but not exactly necessary as dwm provides its own bar and doesn't announce anything like tags via fifo/stdout like bspwm/monsterwm do)

**Comment by bbenne10** [comment]
I think I posted to the wrong thread last night - but even after the `_NET_WM_STRUT` change, wkline works fine in monsterwm

**Comment by alspore** [comment]
@Lokaltog, the commit you referenced didn't fix the issue.

**Comment by Lokaltog** [comment]
Ok, I've added extra handling of strut_partial, could you please check out if 6c53674b0f802873fe52f4b4bdc482411ce8e05a works?

Edit: Crap, wrong indexes. 6c53674b0f802873fe52f4b4bdc482411ce8e05a should be correct.

**Comment by bbenne10** [comment]
This is an interesting bug. The i3 site mentions that all it cares about is the dock hint, but the EWMH docs say that `_NET_WM_STRUT` **MUST** be set when space is to be reserved. Maybe we check if `_NET_WM_WINDOW_TYPE_DOCK` is supported and set both `_NET_WM_WINDOW_TYPE_DOCK` and `_NET_WM_STRUT` if they are and fall back to override redirect in other cases?

**Comment by Lokaltog** [comment]
IIRC dwm and other non-EWMH-compliant WMs usually support the dock hint (but they don't handle dock windows in a particular way), I'm not sure if it's a reliable way to check if the WM actually cares about the dock hint like i3. I noticed that i3 doesn't actually support `_NET_WM_STRUT`, it _only_ supports `_NET_WM_STRUT_PARTIAL` (which is weird if you ask me), so what I've updated the code to do now is to check for support for both of these properties and set them accordingly, and fall back to override-redirect if neither strut property is supported by the WM. Hopefully that will resolve the issue.
